### PR TITLE
compute streaming api close channel on query parsing errors

### PR DESCRIPTION
**Pull request by chwarwick**
In the streaming endpoint when the parsing of the query errored a nil channel was returned which was blocking and caused the endpoint to hang. 

closes https://github.com/sourcegraph/sourcegraph/issues/39835
## Test plan

Without a pattern
```
GET 'https://sourcegraph.test:3443/.api/compute/stream?q=select:file count:all patternType:literal repo:^github.com/sourcegraph/sourcegraph$' 
```

```
HTTP/2 500
compute endpoint expects nonempty pattern
```

With a pattern
```
GET 'https://sourcegraph.test:3443/.api/compute/stream' --data-urlencode 'q=select:file count:all patternType:literal repo:^github.com/sourcegraph/sourcegraph$ hasapattern' 
```

```
event: progress
data: {"done":true,"matchCount":0,"durationMs":9,"skipped":[]}

event: done
data: {}
```

### CODE REVIEW

**Review by camdencheek — APPROVED** [review]
Good catch -- thanks for the fix!

**Review by rvantonder — COMMENTED** [review]
Instead of ensuring this function returns a non-nil stream, can we instead establish a better contract for this function so that the caller can make a decision? Either:

- have this function return an error that the caller can check, or

- check whether the channel returned is `nil` in the caller

The reason is the calling function continues to "do work" when it should be erroring/stopping if the query doesn't parse, here: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@5b7d9c0a0a42e054a5df8a9f9a3a1dae805497f9/-/blob/enterprise/cmd/frontend/internal/compute/streaming/stream.go?L83-84&subtree=true. 

(this is the only caller right now I think)

**Comment by rvantonder** [comment]
Or, probably better, just factor this block out into the caller:

```
	computeQuery, err := compute.Parse(query)
	if err != nil {
		return nil, func() (*search.Alert, error) { return nil, err }
	}

	searchQuery, err := computeQuery.ToSearchQuery()
	if err != nil {
		return nil, func() (*search.Alert, error) { return nil, err }
	}
```

so we don't create a new stream if parsing fails.

**Comment by chwarwick** [comment]
> Or, probably better, just factor this block out into the caller:
> 
> ```go
> 	computeQuery, err := compute.Parse(query)
> 	if err != nil {
> 		return nil, func() (*search.Alert, error) { return nil, err }
> 	}
> 
> 	searchQuery, err := computeQuery.ToSearchQuery()
> 	if err != nil {
> 		return nil, func() (*search.Alert, error) { return nil, err }
> 	}
> ```
> 
> so we don't create a new stream if parsing fails.

@rvantonder Pulled this logic out, I had to update the function signature because it needs the search query and the compute command, that confirmed it's the only caller.

File: enterprise/cmd/frontend/internal/compute/streaming/stream.go [hunk 1]
```diff
@@ -80,7 +93,7 @@ func (h *streamHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	// Log events to trace
 	eventWriter.StatHook = eventStreamOTHook(tr.LogFields)
 
-	events, getResults := NewComputeStream(ctx, h.logger, h.db, args.Query)
+	events, getResults := NewComputeStream(ctx, h.logger, h.db, searchQuery, computeQuery.Command)
```
Comment (rvantonder) [inline enterprise/cmd/frontend/internal/compute/streaming/stream.go:96]: `searchQuery` is unused, no?

**Review by rvantonder — APPROVED** [review]
Great, one minor comment. Thanks for the change! This looks cleaner :-)

File: enterprise/cmd/frontend/internal/compute/streaming/stream.go (hunk 1 above)
Comment (rvantonder) [reply enterprise/cmd/frontend/internal/compute/streaming/stream.go:96]: 🤦 missed that. thanks!

**Pull request merged**